Archive/EulerFermat.lean

```diff
@@ -0,0 +1,169 @@
+/-
+Copyright (c) 2024 Ralf Stephan. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Ralf Stephan
+-/
+import Mathlib
+
+/-!
+# a^n+1 is prime only if n is a power of two
+
+The fact that primes of the form `a^n+1` must have `n` a power of two was stated and
+the proof outlined in three sentences by Euler in the year 1738 [E026]. A proof can be
+found in Hardy & Wright, An introduction to the theory of numbers (5th ed., 1979),
+p.15, Theorem 17. In the proof below we followed Euler's thoughts.
+-/
+
+open Nat
+
+lemma simplify1 (a m' : ℕ) (ha : a > 0) :
+    a ^ (2 * m' + 1) + 1 + (a + 1) * (a ^ (2 * m' + 2) - a ^ (2 * m' + 1)) =
+    a ^ (2 * m' + 3) + 1 :=
+  have h₁ (a b c d : ℕ): a + b + c + d - a = b + c + d := by omega
+  have h₂ (a b : ℕ) (h : b ≤ a): 1 + a - b + b = 1 + a := by omega
+  calc
+    a ^ (2 * m' + 1) + 1 + (a + 1) * (a ^ (2 * m' + 2) - a ^ (2 * m' + 1)) =
+      a ^ (2 * m' + 1) + 1 + a * (a ^ (2 * m' + 2) - a ^ (2 * m' + 1)) +
+      (a ^ (2 * m' + 2) - a ^ (2 * m' + 1)) := by
+      linarith
+    _ = a ^ (2 * m' + 1) + 1 + a * (a ^ (2 * m' + 2) - a ^ (2 * m' + 1)) +
+      a ^ (2 * m' + 2) - a ^ (2 * m' + 1) := by
+      rw [← Nat.add_sub_assoc]
+      exact pow_le_pow_of_le_right ha (le_succ (2 * m' + 1))
+    _ = 1 + a * (a ^ (2 * m' + 2) - a ^ (2 * m' + 1)) + a ^ (2 * m' + 2) := by
+      rw [h₁]
+    _ = 1 + (a * a ^ (2 * m' + 2) - a * a ^ (2 * m' + 1)) + a ^ (2 * m' + 2) := by
+      rw [Nat.mul_sub_left_distrib a (a ^ (2 * m' + 2)) (a ^ (2 * m' + 1))]
+    _ = 1 + (a ^ (2 * m' + 2 + 1) - a ^ (2 * m' + 1 + 1)) + a ^ (2 * m' + 2) := by
+      simp_rw [← Nat.pow_succ']
+    _ = 1 + (a ^ (2 * m' + 3) - a ^ (2 * m' + 2)) + a ^ (2 * m' + 2) := by
+      linarith
+    _ = 1 + a ^ (2 * m' + 3) - a ^ (2 * m' + 2) + a ^ (2 * m' + 2) := by
+      rw [← Nat.add_sub_assoc]
+      exact pow_le_pow_of_le_right ha (le_succ (2 * m' + 2))
+    _ = 1 + a ^ (2 * m' + 3) := by
+      apply h₂ (a ^ (2 * m' + 3)) (a ^ (2 * m' + 2))
+      exact pow_le_pow_of_le_right ha (le_succ (2 * m' + 2))
+    _ = a ^ (2 * m' + 3) + 1 := by
+      exact Nat.add_comm 1 (a ^ (2 * m' + 3))
+
+/-- "$a^{2m+1}+1$ can be divided by $a+1$" [Euler, E026]. Proof by induction. -/
+theorem H1 (a m : ℕ) (ha : a > 0) : (a + 1) ∣ a ^ (2 * m + 1) + 1 := by
+  have h₃ (a m' : ℕ) : a ^ (2 * m' + 2 + 1) = a ^ (2 * m' + 3) := by ring_nf
+  have hdvd (a m' : ℕ) : (a + 1) ∣ (a + 1) * (a ^ (2 * m' + 2) - a ^ (2 * m' + 1)) :=
+    dvd_mul_right (a + 1) (a ^ (2 * m' + 2) - a ^ (2 * m' + 1))
+  induction m with
+  | zero =>
+    exact Dvd.intro_left (Nat.pow a 0) rfl
```

Comment:
You should not need to write `Nat.pow` directly; use `a ^ 0` instead. Also, this can likely be replaced by [`dvd_mul_left`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Divisibility/Basic.html#dvd_mul_left)

Reply:
Done.